src/rhsmlib/facts/hwprobe.py

```diff
@@ -554,21 +558,37 @@ def get_cpu_info(self):
         return cpu_info
 
     def get_ls_cpu_info(self):
-        lscpu_info = {}
-
-        LSCPU_CMD = '/usr/bin/lscpu'
-
         # if we have `lscpu`, let's use it for facts as well, under
         # the `lscpu` name space
-        if not os.access(LSCPU_CMD, os.R_OK):
-            return lscpu_info
+        if not os.access(self.LSCPU_CMD, os.R_OK):
+            return {}
 
         # copy of parent process environment
         lscpu_env = dict(os.environ)
 
         # # LANGUAGE trumps LC_ALL, LC_CTYPE, LANG. See rhbz#1225435, rhbz#1450210
         lscpu_env.update({'LANGUAGE': 'en_US.UTF-8'})
-        lscpu_cmd = [LSCPU_CMD]
+
+        if self._check_lscpu_json(lscpu_env):
+            return self._parse_lscpu_json_output(lscpu_env)
+
+        return self._parse_lscpu_human_readable_output(lscpu_env)
+
+    def _check_lscpu_json(self, lscpu_env):
+        lscpu_cmd = [self.LSCPU_CMD, '--help']
+
+        try:
+            output = subprocess.check_output(lscpu_cmd, env=lscpu_env)
+        except CalledProcessError as e:
+            log.exception(e)
+            log.warning('Error with lscpu --json subprocess: %s', e)
```

Comment:
I am concerned that this will confuse the user. We are logging an 'Error' at warning level.

Perhaps we can change this warn message to 'Unable to retrieve lscpu via subprocess: %s', e) and then remove the logging of the exception. Otherwise, we will get a BZ for it when it is considered a reasonable outcome of the call. Warning feels correct as it does not stop the execution.

Reply:
Right, 'Error' for a warning can be indeed confusing; OTOH, I mostly followed what the current non-JSON lscpu code path does. Should we adapt that too?

I tweaked the warning messages a little bit, IMHO they should be better now. Also, since we print the exception as string, most probably you are right that we can avoid the whole stack trace which, at least for these cases, does not provide an added value. WDYT?